### ISSUE TITLE
Number Token Loss

### ISSUE THREAD

**Issue by jannisborn** [huggingface/transformers#38950]
### Feature request

**Summary:** We suggest to integrate a novel loss, called **Number Token Loss**, to appear at [ICML 2025](https://ibm.biz/ntl-paper).
NTL computes a regression-like loss directly on the token level of a LMHead. **Experiments show that CE augmented with NTL forms a significantly better pretraining objective for LLMs than standard CE.**

**Details**: NTL improves performance on tasks where the tokens are actually numbers, e.g., math datasets. It is rapid to calculate and intended to augment cross entropy in order to provide a more meaningful loss signal if the label is a numerical token.  
NTL-WAS minimizes the Wasserstein-1 distance between the *numerical values* of the real and predicted number tokens. In practice, the Wasserstein-1 distance can be calculated extremely efficiently because the label distribution is one-hot. So no need for any approximations (Sinkhorn distance etc). However, the loss class needs a one-time access to the tokenizer in order to extract a map between tokens (strings) and their numerical value (floats). Also note that this loss does not affect any token that is *not* a number.

### Motivation

Cross entropy is the ubiquitous objective in LMs. But when LMs are trained on text with numbers (e.g., math tasks) there is a fundamental problem: CE is nominal-scale -- it assumes tokens are unordered. That makes sense for words, but not for numbers. For example, if the true label is “5,” predicting “6” or “9” yields the same loss, see demo below.

Next month at ICML, we will present our paper on the [Number Token Loss](https://ibm.biz/ntl-paper) which fixes this. The loss comes in two flavors, NTL-WAS and NTL-MSE. See the demo below.

![Image](https://github.com/user-attachments/assets/97b8a28d-dd51-4e03-a7c1-8ad26a4afda2)

For more details see [here](https://www.linkedin.com/feed/update/urn:li:activity:7340864033874472962/) and [here](https://ibm.biz/ntl-main)

### Your contribution

We would be happy to work toward the integration of NTL into the library, ideally to be exposed via `LOSS_MAPPING` in `transformers.loss.loss_utils`. Would you be willing to accept such a PR?

**Comment by Rocketknight1** [comment]
Hi @jannisborn, the loss looks cool, but the focus of Transformers is generally on supporting pre-trained models, rather than training with novel architectures/losses! Losses usually only go into `loss_utils.py` when major model checkpoints have been trained using them, and we need the loss code to support those models.
